### PR TITLE
Forward kinematics and Arithmetic AST Evaluation

### PR DESCRIPTION
**Pull request by dmorejon**
Implementing forward functions for forward kinematics given a array of angles and segment lengths and forward arithmetic given an AST.

### CODE REVIEW

File: ExampleModels/Arithmetic/src/arithmetic.jl [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    DIV = 4
+end
+
+struct BinaryOp <: Expression
+    left::Expression
+    op::OP
+    right::Expression
+end
+
+function evaluate(expr::Constant)
```
Comment (zenna) [inline ExampleModels/Arithmetic/src/arithmetic.jl:22]: Write very short functions on one line like:
`evaluate(expr::Constant) = expr.value`

File: ExampleModels/Arithmetic/src/arithmetic.jl [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    if expr.op == ADD
+        return leftVal + rightVal
+    elseif expr.op == SUB
+        return leftVal - rightVal
+    elseif expr.op == MUL
+        return leftVal * rightVal
+    elseif expr.op == DIV
+        return leftVal / rightVal
+    else 
+        println("Unknown op while evaluating BinaryOp")
```
Comment (zenna) [inline ExampleModels/Arithmetic/src/arithmetic.jl:38]: throw an exception, instead of `println`

File: ExampleModels/Arithmetic/src/arithmetic.jl [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    end
+end
+
+#=
+
+# Simple Test
```
Comment (zenna) [inline ExampleModels/Arithmetic/src/arithmetic.jl:44]: add tests into `test/runtest.jl` and remove commented code

File: ExampleModels/Arithmetic/src/arithmetic.jl [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    op::OP
+    right::Expression
+end
+
+function evaluate(expr::Constant)
+    expr.value
+end
+
+function evaluate(expr::BinaryOp) 
+    leftVal = evaluate(expr.left)
```
Comment (zenna) [inline ExampleModels/Arithmetic/src/arithmetic.jl:27]: In Julia by convention we use camelCase only for types.  Use `leftval` or `left_val`

File: ExampleModels/Arithmetic/src/arithmetic.jl [hunk 5]
```diff
@@ -0,0 +1,107 @@
+module Arithmetic
+
+abstract type Expression end
+
+struct Constant <: Expression
+    value::Number
```
Comment (zenna) [inline ExampleModels/Arithmetic/src/arithmetic.jl:6]: Please use 2 spaces for indentation (this is just the preference that I like to use for my code bases, not a general rule)

File: ExampleModels/ForwardKinematics/src/forwardKinematics.jl [hunk 6]
```diff
@@ -0,0 +1,37 @@
+module ForwardKinematics
+
```
Comment (zenna) [inline ExampleModels/ForwardKinematics/src/forwardKinematics.jl:2]: rename this file ForwardKinematics.jl (capitalize first letter)

File: ExampleModels/ForwardKinematics/src/forwardKinematics.jl [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    x::Float64
+    y::Float64
+end
+
+# angles is an array of angles, in order from base to end, in radians. 
+# lengths is an array of segment lengths, in order from base to end, in any (consistent) unit.
+# returns Point(x, y) of final segment. 
+# function forwardKinematics(angles::Array{Number}, lengths::Array{Number})
+function forwardKinematics(angles, lengths)
+    @assert (size(angles, 1) == size(lengths, 1)) "unequal parameter sizes"
```
Comment (zenna) [inline ExampleModels/ForwardKinematics/src/forwardKinematics.jl:13]: This shouldn't be an assertion, it should throw an exception.  If you're not clear on when you should assertions vs throw exceptions then look it up or ask me, but in short, @assertions should pick up invariants in your code that should only be violated if your program is incorrect

File: ExampleModels/ForwardKinematics/src/forwardKinematics.jl [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    for i = 1:size(angles, 1)
+        angleSum += angles[i]
+        x += lengths[i] * cos(angleSum)
+        y += lengths[i] * sin(angleSum)
+    end
+    Point(x, y)
+end
+
+#=
+# Simple test
```
Comment (zenna) [inline ExampleModels/ForwardKinematics/src/forwardKinematics.jl:26]: Put these tests into runtests.jl

**Review by zenna — CHANGES_REQUESTED** [review]

File: ExampleModels/ForwardKinematics/src/forwardKinematics.jl (hunk 8 above)
Comment (dmorejon) [reply ExampleModels/ForwardKinematics/src/forwardKinematics.jl]: Updated but can't get the runtests.jl to recognize neither the ForwardKinematics module nor Arithmetic module.

File: ExampleModels/ForwardKinematics/test/runtests.jl [hunk 9]
```diff
@@ -0,0 +1,21 @@
+using ForwardKinematics
+using Test
+
+floatEquals(x, y, eps) = abs(x - y) <= eps
```
Comment (zenna) [inline ExampleModels/ForwardKinematics/test/runtests.jl:4]: use camelCase only for types

Comment (zenna) [reply ExampleModels/ForwardKinematics/test/runtests.jl:4]: Also there is both `isapprox` for testing approximate equality

and also:

```
 julia> @test π ≈ 3.14 atol=0.01
  Test Passed

  This is equivalent to the uglier test @test ≈(π, 3.14, atol=0.01). It is an error to supply more
  than one expression unless the first is a call expression and the rest are assignments (k=v).
```

File: ExampleModels/ForwardKinematics/test/runtests.jl [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+# Simple test
+point = forward_kinematics([pi], [1.0])
+@test floatEquals(point.x, -1, 0.00001)
+@test floatEquals(point.y,  0, 0.00001)
+
+# Exception due to unequal sizes
+try
```
Comment (zenna) [inline ExampleModels/ForwardKinematics/test/runtests.jl:12]: What is going on here?

**Review by zenna — CHANGES_REQUESTED** [review]
A few minor changes, and also put your tests inside `@testset`s

File: ExampleModels/ForwardKinematics/test/runtests.jl (hunk 10 above)
Comment (dmorejon) [reply ExampleModels/ForwardKinematics/test/runtests.jl:12]: This is testing that we properly fail (throw an exception) if the argument have unequal sizes

**Pull request merged**